File: java/2-IsPalindrome.java
// Reversing only half of the number to check for palindrome

public class Solution {
    public boolean isPalindrome(int x) {
        // Edge Case 1: Negative numbers are not palindromes (e.g., -121 becomes 121-)
        // Edge Case 2: If the last digit is 0, the first digit must also be 0 for it to be a palindrome.
        // Only the number 0 itself satisfies this condition.
        if (x < 0 || (x % 10 == 0 && x != 0)) {
            return false;
        }

        int revertedNumber = 0;
        
        // Loop until we reach the middle of the number.
        // When 'x' becomes less than or equal to 'revertedNumber', 
        // it means we have processed half of the digits.
        while (x > revertedNumber) {
            int pop = x % 10;          // Extract the last digit
            revertedNumber = (revertedNumber * 10) + pop; // Push it to the reverted number
            x /= 10;                   // Remove the last digit from x
        }

        // For even-length numbers (e.g., 1221): x will be 12, revertedNumber will be 12.
        // For odd-length numbers (e.g., 12321): x will be 12, revertedNumber will be 123.
        // We can safely get rid of the middle digit by doing revertedNumber / 10.
        return x == revertedNumber || x == revertedNumber / 10;
    }
}

// Logic Explanation

// To determine if a number is a palindrome without using extra string space, we process the digits from right to left and build a new number. However, to avoid integer overflow issues, we only reverse half of the number.

// 1. Filtering Edge Cases
// Negative Numbers: Any negative number (like -121) cannot be a palindrome because the minus sign reads at the start but would be at the end if reversed.

// Trailing Zeros: Any number ending in 0 (like 10, 290) cannot be a palindrome because no standard integer starts with a 0, except for 0 itself.

// 2. Finding the Middle Point
// We strip the last digit of x using x % 10 and append it to revertedNumber. Then, we divide x by 10.

// How do we know we've hit the middle? Since we are moving digits from x to revertedNumber, once x <= revertedNumber, we have officially reached or passed the halfway mark.

// 3. The Final Comparison
// Even Number of Digits (e.g., 1221): At the end of the loop, x = 12 and revertedNumber = 12. They match exactly.

// Odd Number of Digits (e.g., 12321): At the end of the loop, x = 12 and revertedNumber = 123. The middle digit (3) doesn't affect palindrome status, so we truncate it using revertedNumber / 10 (which gives 12) and check if it matches x.

File: python/easy/2_is_palindrome.py
class Solution:
    def isPalindrome(self, x: int) -> bool:
        # Special cases:
        # As discussed above, when x < 0, x is not a palindrome.
        # Also if the last digit of the number is 0, the first digit of the number also needs to be 0.
        # Only 0 satisfies this property.
        if x < 0 or (x % 10 == 0 and x != 0):
            return False

        reverted_number = 0
        while x > reverted_number:
            # Get the last digit of x and append it to reverted_number
            reverted_number = (reverted_number * 10) + (x % 10)
            # Remove the last digit from x
            x //= 10

        # When the length is an odd number, we can get rid of the middle digit by reverted_number // 10
        # For example, when x = 121, at the end of the while loop we get x = 1, reverted_number = 12
        # Since the middle digit doesn't matter in palindrome, we can safely remove it.
        return x == reverted_number or x == reverted_number // 10
    
# Algorithm Used: Reversing Half of the Number

# Edge Cases
# - All negative numbers are definitely not palindromes due to the leading minus sign (e.g., -121).
# - If the number ends with 0, it can only be a palindrome if the number itself is 0 (e.g., 10 or 200 are not palindromes).

# Reversing Process
# 1. Initialize a variable reverted_number = 0.

# 2. Loop as long as x > reverted_number. Inside the loop:
# -- Extract the last digit of x using the modulo operator (x % 10) and append it to reverted_number.
# -- Truncate the last digit from x using integer division (x // 10).    